src/main/java/slimeknights/tconstruct/library/tinkering/ToolPartItem.java

```diff
@@ -0,0 +1,170 @@
+    boolean shift = Util.isShiftKeyDown();
+
+    // todo traits
+    //if(!this.checkMissingMaterialTooltip(stack, tooltip)) {
+    //  tooltip.addAll(getTooltipTraitInfo(material));
+    //}
+
+    // Stats
+    if (Config.CLIENT.extraToolTips.get()) {
+      if (!shift) {
+        // info tooltip for detailed and component info
+        tooltip.add(new StringTextComponent(""));
+        tooltip.add(new TranslationTextComponent("tooltip.tool.hold_shift"));
+      }
+      else {
+        tooltip.addAll(this.getTooltipStatsInfo(material));
+      }
+    }
+
+    tooltip.addAll(this.getAddedByInfo(material));
+  }
+
+  public List<ITextComponent> getTooltipStatsInfo(IMaterial material) {
+    ImmutableList.Builder<ITextComponent> builder = ImmutableList.builder();
+
+    Optional<IMaterialStats> materialStat = MaterialRegistry.getInstance().getMaterialStats(material.getIdentifier(), this.materialStatId);
+    if (materialStat.isPresent()) {
+      IMaterialStats stat = materialStat.get();
+
+      if (this.hasUseForStat(stat.getIdentifier())) {
+        List<ITextComponent> text = stat.getLocalizedInfo();
+        if (!text.isEmpty()) {
+          builder.add(new StringTextComponent(""));
+          builder.add(stat.getLocalizedName().applyTextStyles(TextFormatting.WHITE, TextFormatting.UNDERLINE));
+          builder.addAll(stat.getLocalizedInfo());
+        }
+      }
+    }
+
+    return builder.build();
+  }
+
+  public List<ITextComponent> getAddedByInfo(IMaterial material) {
+    ImmutableList.Builder<ITextComponent> builder = ImmutableList.builder();
+
+    if (MaterialRegistry.getInstance().getMaterial(material.getIdentifier()) != IMaterial.UNKNOWN) {
+      builder.add(new StringTextComponent(""));
+      for (ModInfo modInfo : ModList.get().getMods()) {
+        if (modInfo.getModId().equalsIgnoreCase(material.getIdentifier().getNamespace())) {
+          builder.add(new TranslationTextComponent("tooltip.part.material_added_by", modInfo.getDisplayName()));
+        }
+      }
+    }
+
+    return builder.build();
+  }
+
+  public boolean checkMissingMaterialTooltip(ItemStack stack, List<String> tooltip) {
+    return checkMissingMaterialTooltip(stack, tooltip, "");
+  }
+
+  public boolean checkMissingMaterialTooltip(ItemStack stack, List<String> tooltip, String statIdentifier) {
+    IMaterial material = this.getMaterial(stack);
+
+    if (material == IMaterial.UNKNOWN) {
+      CompoundNBT tagSafe = TagUtil.getTagSafe(stack);
+      String materialId = tagSafe.getString(Tags.PART_MATERIAL);
+      ITextComponent error;
+
+      if (!materialId.isEmpty()) {
+        error = new TranslationTextComponent("tooltip.part.missing_material", materialId);
+      }
+      else {
+        error = new TranslationTextComponent("tooltip.part.missing_info");
+      }
+      tooltip.add(error.getFormattedText());
+      return true;
+    }
+    /*else if(statIdentifier != null && material.getStats(statIdentifier) == null) {
```

Comment:
Why is this commented out? It looks like it would be useful

Reply:
`getStats(statIdentifier)` is not a method on IMaterial, so I commented it out